tests/Freya.Integration.Tests/Integration.Tests.fs

```diff
@@ -0,0 +1,278 @@
+    unbox env.["o2"] =? false
+    unbox env.["Answer"] =? "42,1"
+
+[<Test>]
+let ``pipeline executes both monads if first returns next with wrapped pipeline OwinMidFunc`` () =
+    let o1 = o1 *> next
+    let o2 = o2 *> next
+    let midFunc = OwinMidFunc.ofFreyaWrapped o1 o2
+    let composed = midFunc.Invoke app
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? true
+    unbox env.["Answer"] =? "1,42,2"
+
+[<Test>]
+let ``pipeline executes only the first monad if first returns terminate with wrapped pipeline OwinMidFunc`` () =
+    let o1 = o1 *> halt
+    let o2 = o2 *> next
+    let midFunc = OwinMidFunc.ofFreyaWrapped o1 o2
+    let composed = midFunc.Invoke app
+
+    let env = invoke composed
+
+    unbox env.["o1"] =? true
+    unbox env.["o2"] =? false
+    unbox env.["Answer"] =? "1"
+
+[<Test>]
+let ``MidFunc can roundtrip to Freya Pipeline before / after and back`` () =
+    let midFunc =
+        OwinMidFunc(fun next ->
+            OwinAppFunc(fun env ->
+                async {
+                    env.["o3"] <- true
+                    do! next.Invoke(env).ContinueWith<unit>(fun _ -> ()) |> Async.AwaitTask
+                    env.["o4"] <- true
+                } |> Async.StartAsTask :> Task ))
+
+    let before, after = OwinMidFunc.splitIntoFreya midFunc
+    let pipe = OwinMidFunc.ofFreyaWrapped before after
+    let composed = pipe.Invoke app
+
+    let env = invoke composed
+
+    env.ContainsKey("Answer") =? true
+    unbox env.["Answer"] =? "42"
+    env.ContainsKey("o3") =? true
+    unbox env.["o3"] =? true
+    env.ContainsKey("o4") =? true
+    unbox env.["o4"] =? true
+
+[<Test>]
+let ``MidFunc that terminates early and doesn't call next Halts correctly as a Freya Pipeline`` () =
+    let midFunc =
+        OwinMidFunc(fun next ->
+            OwinAppFunc(fun env ->
+                env.["o3"] <- true
+                // Don't call next
+                Task.FromResult<obj>(null) :> Task ))
+
+    let before, after = OwinMidFunc.splitIntoFreya midFunc
+    let pipe = OwinMidFunc.ofFreyaWrapped before after
+    let composed = pipe.Invoke app
+
+    let env = invoke composed
+
+    env.ContainsKey("o3") =? true
+    unbox env.["o3"] =? true
+    env.ContainsKey("Answer") =? false
+    env.ContainsKey("o4") =? false
+
+[<Test>]
+let ``MidFunc that throws early and doesn't call next Halts correctly as a Freya Pipeline`` () =
+    let midFunc =
+        OwinMidFunc(fun next ->
+            OwinAppFunc(fun env ->
+                env.["o3"] <- true
+                // Don't call next
+                let tcs = TaskCompletionSource<unit>()
+                tcs.SetException(System.Exception("failed!"))
+                tcs.Task :> Task ))
+
+    let before, after = OwinMidFunc.splitIntoFreya midFunc
+    let pipe = OwinMidFunc.ofFreyaWrapped before after
+    let composed = pipe.Invoke app
+
+    let env = invoke composed
+
+    // TOOD: What *should* happen here?
```

Comment:
Very important `TODO`: I don't know what is actually supposed to happen with a faulted `Task`.